ark/jsonschema/object.ts

```diff
@@ -0,0 +1,265 @@
+
+	if (
+		"domain" in propertyNamesValidator.json &&
+		propertyNamesValidator.json.domain !== "string"
+	) {
+		ctx.reject({
+			path: ["propertyNames"],
+			expected: "a schema for validating a string",
+			actual: `a schema for validating a ${printable(propertyNamesValidator.json.domain)}`
+		})
+	}
+
+	const jsonSchemaObjectPropertyNamesValidator = (
+		data: object,
+		ctx: Traversal
+	) => {
+		Object.keys(data).forEach(key => {
+			if (!propertyNamesValidator.allows(key)) {
+				ctx.reject({
+					path: [key],
+					expected: `a key adhering to the propertyNames schema of ${propertyNamesValidator.description}`,
+					actual: key
+				})
+			}
+		})
+		return !ctx.hasError()
+	}
+	return jsonSchemaObjectPropertyNamesValidator
+}
+
+const parseRequiredAndOptionalKeys = (
+	jsonSchema: JsonSchema.Object,
+	ctx: Traversal
+) => {
+	const optionalKeys: string[] = []
+	const requiredKeys: string[] = []
+	if ("properties" in jsonSchema) {
+		if ("required" in jsonSchema) {
+			const duplicateRequiredKeys = getDuplicatesOf(jsonSchema.required)
+			if (duplicateRequiredKeys.length !== 0) {
+				ctx.reject({
+					path: ["required"],
+					expected: "an array of unique strings",
+					actual: `an array with the following duplicates: ${printable(duplicateRequiredKeys)}`
+				})
+			}
+
+			for (const key of jsonSchema.required) {
+				if (key in jsonSchema.properties) requiredKeys.push(key)
+				else {
+					ctx.reject({
+						path: ["required"],
+						expected: `a key from the 'properties' object (one of ${printable(Object.keys(jsonSchema.properties))})`,
+						actual: key
+					})
+				}
+			}
+			for (const key in jsonSchema.properties)
+				if (!jsonSchema.required.includes(key)) optionalKeys.push(key)
+		} else {
+			// If 'required' is not present, all keys are optional
+			optionalKeys.push(...Object.keys(jsonSchema.properties))
+		}
+	} else if ("required" in jsonSchema) {
+		ctx.reject({
+			expected: "a valid object JSON Schema",
+			actual:
+				"an object JSON Schema with 'required' array but no 'properties' object"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: parseJsonSchema(jsonSchema.properties![key]).internal
+		})),
+		requiredKeys: requiredKeys.map(key => ({
+			key,
+			value: parseJsonSchema(jsonSchema.properties![key]).internal
+		}))
+	}
+}
+
+const parseAdditionalProperties = (jsonSchema: JsonSchema.Object) => {
+	if (!("additionalProperties" in jsonSchema)) return
+
+	const properties =
+		jsonSchema.properties ? Object.keys(jsonSchema.properties) : []
+	const patternProperties = Object.keys(jsonSchema.patternProperties ?? {})
+
+	const additionalPropertiesSchema = jsonSchema.additionalProperties
+	if (additionalPropertiesSchema === true) return true
+	if (additionalPropertiesSchema === false) return false
+
+	const jsonSchemaObjectAdditionalPropertiesValidator = (
+		data: object,
+		ctx: Traversal
+	) => {
+		Object.keys(data).forEach(key => {
```

Comment:
It feels wrong to be referring to the structure of the original JSON schema in a closure we're using for validation.

Maybe a more arktype-y way to do this would be to write a narrow like `type.string.narrow(s =>!unionOfExistingKeys.allows(s))`.

Reply:
I'm not quite sure what this comment is referring to, but maybe this is what you meant? f4a6abbbb

Reply:
Yeah this looks good for now. Maybe at some point there's a way to make it more serializable but at the moment this seems like the best solution.